training script modification

samuelmyoussef opened the pull request:
Modified the `rl.py` script:

- Removed algorithms from stable baselines (SB2)
- Added the remaining algorithms from stable baselines (SB3): A2C, DQN, TD3, DDPG
- Added all the available environments in SofaGym
- Added command line arguments to choose the env, algo, and training parameters
- Added the choice to continue training from a previously trained model
- Added Tensorboard logging

Added a documentation part to briefly explain the script usage.

EtienneMenager commented on rl.py:
```diff
@@ -182,71 +320,133 @@ def sec_to_hours(seconds):
+        print(">>   End.")
+        print("[INFO]  >>    time: ", sec_to_hours(time.time()-start_time))
+    
+    elif TRAIN == 'continue':
```
Nice to add the possibility to continue the training. Do you have a specific initialisation of the *idx* and *best* parameters?

samuelmyoussef replied:
You can choose the saved time step you want to continue from with the command line argument "-ms", but the default is the latest time step in the case of continuing the training, and the best in the case of testing the model.

I still have a problem reloading the idx from where it previously stopped. I could save it as a parameter in a separate file alongside maybe the model hyperparameters and load it back with the model, but for now it restarts from 0, but in the tensorboard logging it will show the total timesteps.

EtienneMenager replied:
Ok perfect!

alxbilger commented on rl.py:
```diff
@@ -32,144 +33,281 @@
+videos_dir = f"{logs_dir}/videos"
+
+envs = {
+        1: 'bubblemotion-v0',
```
Why `v0` at the end of all names?

samuelmyoussef replied:
It's a convention in gym to add the env version after the name, in this case version 0. If a significant change is made to the environment like changing how its reward is calculated for example, it would be appropriate to change it to v1.

alxbilger commented on README.md:
````diff
@@ -103,6 +103,28 @@ The classic running of an episode is therefore:
+```
+
+Command line arguments can be used to change the training parameters:
+- -e, --env: environment name from available examples
````
It could be great to be able to list the names of all available examples

samuelmyoussef replied:
linked a reference to the environments section where they are already listed
https://github.com/SofaDefrost/SofaGym/blob/a02a705c4d3b4fa88765b48c6ed5c0c602860349/README.md?plain=1#L114

alxbilger commented on README.md:
```diff
@@ -103,6 +103,28 @@ The classic running of an episode is therefore:
+
+Command line arguments can be used to change the training parameters:
+- -e, --env: environment name from available examples
+- -a, --algo: RL algorithm name from the algorithms supported by stable baselines
```
It could be great to be able to list the names of all available algorithms

samuelmyoussef replied:
included them
https://github.com/SofaDefrost/SofaGym/blob/a02a705c4d3b4fa88765b48c6ed5c0c602860349/README.md?plain=1#L115-L118

alxbilger commented on README.md:
```diff
@@ -103,6 +103,28 @@ The classic running of an episode is therefore:
+- -tn, --num_test: number of tests to run on the trained model [Optional, default=1]
+- -md, --model_dir: path to save the model training files
+- -ms, --model_step: time step to load a trained model from
+
```
I see some parameters from the script are missing from the documentation

samuelmyoussef replied:
added all the parameters in the documentation

alxbilger commented on rl.py:
```diff
@@ -32,144 +33,281 @@
+    TEST = args.test
+    n_test = args.num_test
+
+    layers_size = [512, 512, 512]
```
It is usually not a good practice to use hard coded values. But it is maybe out of the scope of this PR

samuelmyoussef replied:
I moved this part to yaml files to save all the parameters of the model and the training, they can be changed by editing the yaml files directly

alxbilger commented on rl.py:
```diff
@@ -32,144 +33,281 @@
-    value_coeff = value_coeffs[args.num_env]
-    batch_size = batch_sizes[args.num_env]
-    size_layer = size_layers[args.num_env]
+    if algo_name == 'SAC':
```
Polymorphism must be used instead of multiple conditions

alxbilger commented on rl.py:
```diff
@@ -182,71 +320,133 @@ def sec_to_hours(seconds):
-                   json.dump([rewards, steps], fp)
-               with open("./Results_benchmark/" +  name + "/final_rewards_"+id+".txt", 'w') as fp:
-                   json.dump([final_rewards, steps], fp)
+    if TRAIN == 'new':
```
Again polymorphism must be used. At minimum, it must be separated in 2 functions.

alxbilger reviewed (changes requested):
Overall the script becomes too long, therefore difficult to read. Polymorphism could be a option to refactor parts of the script and improve the readability. Do not hesitate to separate the script into several.

samuelmyoussef commented:
I made modifications according to @alxbilger recommendations:
- Implemented classes for the RL agent to separate them according to the framework used:
  - SofaTestAgent: to create an agent only to test the env
  - SofaBaseAgent: the base class for the training classes
  - SB3Agent: to train the agent using Stable Baselines3
  - RlberryAgent: to train the agent using rlberry
- Added all model hyperparameters and training parameters into yaml files to be edited directly instead of parsing command line arguments.
- The `rl.py` script could now be used to create and train an RL agent by specifying the wanted environment, the algorithm, and the framework to use (SB3 or rlberry for now), and the interface is the same whichever the framework chosen.

alxbilger reviewed (changes requested):
Could you just document the base classes (SofaBaseAgent and SofaTestAgent)?
